core/src/main/java/org/infinispan/statetransfer/StateTransferLockImpl.java

```diff
@@ -71,38 +88,38 @@ public void notifyTransactionDataReceived(int topologyId) {
          log.tracef("Signalling transaction data received for topology %d", topologyId);
       }
       transactionDataTopologyId = topologyId;
+
       transactionDataLock.lock();
+      CompletableFuture<Void> oldFuture = transactionDataFuture;
       try {
-         transactionDataCondition.signalAll();
+         transactionDataFuture = new CompletableFuture<>();
       } finally {
          transactionDataLock.unlock();
+         remoteExecutor.execute(() -> oldFuture.complete(null));
```

Comment:
Haha, I wanted to publish my own version for comparison but it doesn't look that different now :)

One difference is that I removed the lock completely. I'm not sure if it's safe, I did have some new failures, but at the very least we could have a single lock for topology and for tx data. I'm also not using the remote executor here, because I want each blocked task to run on a different thread.